### PR TITLE
WOT: Ionq skeleton

### PR DESCRIPTION
**Pull request by dabacon**
OK, here is a skeleton (only 13 days too late for Halloween), without tests (WOT wat?) corresponding to RFC #3479.

If this looks good I will add tests, then start expanding it to completion.

### CODE REVIEW

**Comment by balopat** [comment]
> If this looks good I will add tests, then start expanding it to completion.

Looks good to me, let's see those tests! :D

**Comment by dabacon** [comment]
Someone needs to yell at googlebot :)

**Comment by balopat** [comment]
> Someone needs to yell at googlebot :)

You have some commits from `bacon@bacon-laptop.lan` - those need to change and it should work then!

File: cirq/ionq/ionq_client.py [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+        assert url.scheme is not None and url.netloc is not None, (
+            f'Specified remote_host {remote_host} is not a valid url, '
+            'for example http://example.com')
+        assert api_version in self.SUPPORTED_VERSIONS, (
+            f'Only api v0.1 is accepted but was {api_version}')
+        assert default_target in self.SUPPORTED_TARGETS, (
+            f'Target can only be one of  {self.SUPPORTED_TARGETS} but was '
+            f'{default_target}.')
+        assert max_retry_seconds >= 0, (
+            'Negative retry not possible without time machine.')
```
Comment (vtomole) [inline cirq/ionq/ionq_client.py:94]: :laughing:

File: cirq/ionq/ionq_client.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+    ) -> dict:
+        """Create a job.
+
+        Args:
+            json_circuit: A dict corresponding to the json encoding of the
+                IonQ API.
+            repetitions: The number of times to repeat the circuit. Only can
+                be set if the target is 'qpu'.
+            target: If supplied the target to run on. Supports one of 'qpu' or
+                'simulator'. If not set, uses `defaul_target`.
```
Comment (vtomole) [inline cirq/ionq/ionq_client.py:182]:
```suggestion
                'simulator'. If not set, uses `default_target`.
```

File: cirq/ionq/ionq_client.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+            name: An optional name of the job. Different than the `job_id` of
+                the job.
+
+        Returns:
+            The json body of the response. This does not contain popualted
```
Comment (vtomole) [inline cirq/ionq/ionq_client.py:187]:
```suggestion
            The json body of the response. This does not contain populated
```

File: cirq/ionq/job.py [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+        This is the id used for identifying the job by the API.
+        """
+        return self._job['id']
+
+    def status(self) -> str:
+        """Gets the current status of the job.
+
+        This will get a new job if the status of the job previously was
+        determined to not be in a terminal state. A full list of state is
```
Comment (vtomole) [inline cirq/ionq/job.py:77]:
```suggestion
        determined to not be in a terminal state. A full list of states is
```

File: cirq/ionq/job.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+    def repetitions(self) -> Optional[int]:
+        """Returns the number of repetitions for the job.
+
+        If run on the simulator this will return None.
+        """
+        if 'metadata' in self._job and 'shots' in self._job['metadata']:
+            return int(self._job['metadata']['shots'])
+        return None
+
+    def results(self, timeout_seconds=7200, polling_seconds=1
```
Comment (vtomole) [inline cirq/ionq/job.py:112]: Add types to `timeout_seconds` and `polling_seconds`

File: cirq/ionq/results.py [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+"""Result types for the IonQ API."""
+
+from typing import Dict
+
+from cirq.value import digits
+from cirq import study
+
+
+def _little_endian_to_big(value: int, bit_count: int) -> int:
+    return digits.big_endian_bits_to_int(
```
Comment (vtomole) [inline cirq/ionq/results.py:22]: Feels like we should move this to a  `little_endian_to_big` to `cirq.value.digits`. Might be needed by others in the future.

File: cirq/ionq/results.py [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+        The value is the number of times that corresponding bit string
+        occurred.
+
+        See `to_cirq_result` to convert to a `cirq.Result`.
+        """
+        return self._counts
+
+    def to_cirq_result(self) -> study.Result:
+        # TODO: Convert his to cirq result objects.
```
Comment (vtomole) [inline cirq/ionq/results.py:64]: Very small Nit
```suggestion
        # TODO: Convert this to cirq result objects.
```

File: cirq/ionq/results.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+    """The results of running on an IonQ simulator.
+
+    The IonQ simulator returns the probabilities of the different outcomes,
+    not the raw wavefunction or samples.
+    """
+
+    def __init__(self, probabilities: dict, num_qubits: int):
+        # IonQ returns results in little endian, Cirq prefers to use big endian,
+        # so we convert.
+        print(probabilities)
```
Comment (vtomole) [inline cirq/ionq/results.py:81]: Debug print?

File: cirq/ionq/results.py [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+        self._num_qubits = num_qubits
+
+    def num_qubits(self):
+        """Returns the number of qubits the circuit was run on."""
+        return self._num_qubits
+
+    def probabilities(self):
+        """Returns the probabilities of the measurement results.
+
+        These are the probabilites arrising when measuring all qubits in the
```
Comment (vtomole) [inline cirq/ionq/results.py:95]:
```suggestion
        These are the probabilities arising when measuring all qubits in the
```

**Review by vtomole — CHANGES_REQUESTED** [review]

File: cirq/ionq/ionq_client.py (hunk 1 above)
Comment (dabacon) [reply cirq/ionq/ionq_client.py:94]: glad someone noticed

**Comment by dabacon** [comment]
Obviously need to get spell checking turned on in vs code.

**Comment by dabacon** [comment]
This PR is borked.  Will open new one.  I've fixed @vtomole comments.

**Pull request closed without merging**